### PR TITLE
Multi pdport

### PR DESCRIPTION
**Pull request by olbjo**
Almost completes mapping of multiple ports for pdport config files and profinet index read / write. 
What still remains is to handle interface vs ports.

### CODE REVIEW

**Review by pyhys — COMMENTED** [review]
If I search for PNET_SUBMOD and PNET_SUBSLOT I'm also finding stuff in:
```
src/device/pf_block_writer.c
src/device/pf_cmrdr.c
src/common/pf_alarm.c
```
It would great if we also could fix `pf_port_get_list_of_ports()`

File: src/device/pf_pdport.c [hunk 1]
```diff
@@ -154,21 +187,40 @@ static int pf_pdport_save (pnet_t * net)
...
 static uint16_t pf_pdport_loc_port_num_to_dap_subslot (int loc_port_num)
 {
-   CC_ASSERT (loc_port_num == PNET_PORT_1);
-   return PNET_SUBSLOT_DAP_INTERFACE_1_PORT_1_IDENT;
+   uint16_t subslot =
+      PNET_SUBSLOT_DAP_INTERFACE_1_PORT_1_IDENT + PNET_PORT_1 - loc_port_num;
+   return subslot;
+}
+
+static int pf_pdport_dap_subslot_to_local_port (uint16_t subslot)
```
Comment (pyhys) [inline src/device/pf_pdport.c:204]: Add docstring.
Describe that 0 is returned for subslots out of range.

File: src/device/pf_pdport.c [hunk 2]
```diff
@@ -154,21 +187,40 @@ static int pf_pdport_save (pnet_t * net)
...
+{
+   int port = PNET_PORT_1 + PNET_SUBSLOT_DAP_INTERFACE_1_PORT_1_IDENT - subslot;
+   if (port < PNET_PORT_1 || port > PNET_MAX_PORT)
+   {
+      port = 0;
+   }
+   return port;
+}
+
+static bool pf_pdport_subslot_is_dap_port_id (uint16_t subslot)
```
Comment (pyhys) [inline src/device/pf_pdport.c:214]: Add docstring

File: src/device/pf_pdport.c [hunk 3]
```diff
@@ -228,69 +283,92 @@ static void pf_pdport_remove_all_diag (pnet_t * net)
...
 
-void pf_pdport_ar_connected (pnet_t * net, const pf_ar_t * p_ar)
+void pf_pdport_remove_data_files (const char * file_directory)
 {
-   /* Todo check which dap slots are active in ar and map to ports */
-
-   pf_port_t * p_port_data = pf_port_get_state (net, PNET_PORT_1);
-   if (p_port_data->pdport.adjust.active == true)
+   int port;
+   for (port = PNET_PORT_1; port <= PNET_MAX_PORT; port++)
```
Comment (pyhys) [inline src/device/pf_pdport.c:330]: Please use `pf_port_get_next()`

File: src/device/pf_pdport.c [hunk 4]
```diff
@@ -228,69 +283,92 @@ static void pf_pdport_remove_all_diag (pnet_t * net)
...
+   port = pf_port_get_next (&port_iterator);
 
-   pf_port_t * p_port_data = pf_port_get_state (net, PNET_PORT_1);
-   if (p_port_data->pdport.adjust.active == true)
+   while (port != 0)
    {
-      if (
-         p_port_data->pdport.adjust.peer_to_peer_boundary.peer_to_peer_boundary
-            .do_not_send_LLDP_frames == 1)
+      if (p_port_data->pdport.adjust.active == true)
```
Comment (pyhys) [inline src/device/pf_pdport.c:353]: We need to do `pf_port_get_state() `first?

**Review by pyhys — COMMENTED** [review]
Great!
Just some minor details.

File: src/device/pf_pdport.c (hunk 4 above)
Comment (olbjo) [reply src/device/pf_pdport.c:353]: Yes, that was causing the unit test failure.
Done

File: src/device/pf_pdport.c [hunk 5]
```diff
@@ -368,11 +464,11 @@ int pf_pdport_read_ind (
       }
       break;
    case PF_IDX_DEV_PDREAL_DATA:
-      if (
-         (p_read_req->slot_number == PNET_SLOT_DAP_IDENT) &&
-         ((p_read_req->subslot_number == PNET_SLOT_DAP_IDENT) ||
-          (p_read_req->subslot_number == PNET_SUBMOD_DAP_IDENT)))
+      if (((subslot == PNET_SUBSLOT_DAP_WHOLE_MODULE) ||
+           (subslot == PNET_SUBMOD_DAP_IDENT)))
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/device/pf_pdport.c:468]: Done

File: src/device/pf_pdport.c (hunk 3 above)
Comment (olbjo) [reply src/device/pf_pdport.c:330]: This is only used in a special case where calling function do not have "net". 
"net" is a unused parameter in the port iterator interface so I could remove it there but maybe there is an idea that "net" may be useful in the future in the port iterator. I did not dare to remove or starting to add "net" in used callchain.

File: src/device/pf_pdport.c (hunk 1 above)
Comment (olbjo) [reply src/device/pf_pdport.c:204]: Done

File: src/device/pf_pdport.c (hunk 2 above)
Comment (olbjo) [reply src/device/pf_pdport.c:214]: Done

File: src/device/pf_pdport.c (hunk 3 above)
Comment (pyhys) [reply src/device/pf_pdport.c:330]: Fine! Maybe add:
`/* Do not use port iterator as net not is available */`

File: src/common/pf_alarm.c [hunk 6]
```diff
@@ -2731,19 +2731,45 @@ int pf_alarm_send_diagnosis (
...
-      // device ‘b’
+      ret = pf_cmdev_get_module_ident (net, api_id, slot_nbr, &module_ident);
+      if (ret != 0)
+      {
+         LOG_ERROR (
+            PF_ALARM_LOG,
+            "Alarm(%d): Failed to get module ident for slot: %u",
+            __LINE__,
+            slot_nbr);
+      }
```
Comment (pyhys) [inline src/common/pf_alarm.c:2742]: Should we `return ret; `at error? Also below.

File: src/device/pf_pdport.c [hunk 7]
```diff
@@ -228,69 +299,117 @@ static void pf_pdport_remove_all_diag (pnet_t * net)
...
+void pf_pdport_remove_data_files (const char * file_directory)
+{
+   int port;
+   for (port = PNET_PORT_1; port <= PNET_MAX_PORT; port++)
+   {
+      pf_file_clear (file_directory, pf_pdport_get_filename (port));
+   }
+}
+
 void pf_pdport_ar_connected (pnet_t * net, const pf_ar_t * p_ar)
```
Comment (pyhys) [inline src/device/pf_pdport.c:352]: Maybe `pf_pdport_ar_connect_ind()` 
Now it sounds like `bool pf_pdport_is_ar_connected()`
Just a suggestion

**Review by pyhys — COMMENTED** [review]
Just two very quick fixes, then I'm happy.

File: src/device/pf_port.c [hunk 8]
```diff
@@ -43,11 +43,25 @@ void pf_port_init(pnet_t * net)
...
+   pf_lldp_port_list_t * p_list)
 {
-   /* TODO: Implement support for multiple ports */
+   uint16_t port;
+   uint16_t index = 0;
+   uint8_t port_bit_mask = 0;
    memset (p_list, 0, sizeof (*p_list));
-   p_list->ports[0] = BIT (8 - 1);
+
+   CC_ASSERT (sizeof (*p_list) > (max_port / 8));
```
Comment (pyhys) [inline src/device/pf_port.c:56]: To fix unittest:
`CC_ASSERT (sizeof (*p_list) >= (max_port / 8));`

File: src/device/pf_pdport.c (hunk 3 above)
Comment (olbjo) [reply src/device/pf_pdport.c:346]: Done

File: src/device/pf_port.c (hunk 8 above)
Comment (olbjo) [reply src/device/pf_port.c:56]: I moved the assert into the loop to avoid complex condition. 
Done

File: src/common/pf_alarm.c (hunk 6 above)
Comment (olbjo) [reply src/common/pf_alarm.c:2742]: Updated so return code is -1 also if operation fails to get the module ids.
Done

File: src/device/pf_pdport.c (hunk 7 above)
Comment (olbjo) [reply src/device/pf_pdport.c:352]: Changed to pf_pdport_ar_connect_ind()
Done

**Review by pyhys — APPROVED** [review]
Great!

**Pull request merged**